### ISSUE TITLE
[ICE]: `entered unreachable code: unexpected autoderef error [FulfillmentError { .. } `

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust#151324]
auto-reduced (treereduce-rust):
````rust
//@compile-flags: --edition=2024 -Znext-solver=globally
use std::collections::HashMap;

async fn real_main() {
    let mut map: HashMap<(), ()> = real_main();
    map.insert(());
}

````

original:
````rust
//@ edition:2021
//@ build-pass

use std::collections::HashMap;

fn collections() {
    let _ = real_main();
}

async fn nop() {}

async fn real_main() {
    nop().await;
    real_main().await;
    nop().await;
    HashMap::new().await;

    let mut map: HashMap<(), ()> = real_main();
    map.insert((), nop().await);
}

````

Version information
````
rustc 1.95.0-nightly (621d76794 2026-01-18)
binary: rustc
commit-hash: 621d76794c76fc21c0a6151fbc090120e0230a91
commit-date: 2026-01-18
host: x86_64-unknown-linux-gnu
release: 1.95.0-nightly
LLVM version: 21.1.8

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/621d76794c76fc21c0a6151fbc090120e0230a91/compiler/rustc_hir_typeck/src/method/probe.rs#L1901-L1913

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc --edition=2024 -Znext-solver=globally`

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.TRPUTK5reeMi/rustc_testrunner_tmpdir_reporting.JyFHmehc1bP7/mvce.rs:6:2
  |
6 | }
  |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.TRPUTK5reeMi/rustc_testrunner_tmpdir_reporting.JyFHmehc1bP7/mvce.rs`


thread 'rustc' (1787545) panicked at /rustc-dev/621d76794c76fc21c0a6151fbc090120e0230a91/compiler/rustc_hir_typeck/src/method/probe.rs:1907:21:
internal error: entered unreachable code: unexpected autoderef error [FulfillmentError { obligation: Obligation(predicate=Binder { value: TraitPredicate(<std::collections::HashMap<(), ()> as std::future::Future>, polarity:Positive), bound_vars: [] }, depth=2), code: Unimplemented, root_obligation: Obligation(predicate=Binder { value: AliasRelate(Term::Ty(Alias(Opaque, AliasTy { args: [], def_id: DefId(0:5 ~ mvce[a621]::real_main::{opaque#0}), .. })), Equate, Term::Ty(std::collections::HashMap<(), (), std::hash::RandomState, std::alloc::Global>)), bound_vars: [] }, depth=0) }]
stack backtrace:
   0:     0x7f9902433fd3 - <<std[287527946f71f0bc]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[7477a31a1a34cc49]::fmt::Display>::fmt
   1:     0x7f9902a0e2c8 - core[7477a31a1a34cc49]::fmt::write
   2:     0x7f990244a9c6 - <std[287527946f71f0bc]::sys::stdio::unix::Stderr as std[287527946f71f0bc]::io::Write>::write_fmt
   3:     0x7f990240a048 - std[287527946f71f0bc]::panicking::default_hook::{closure#0}
   4:     0x7f9902427703 - std[287527946f71f0bc]::panicking::default_hook
   5:     0x7f99013fbb8a - std[287527946f71f0bc]::panicking::update_hook::<alloc[979041e9952e771b]::boxed::Box<rustc_driver_impl[517260a07841a630]::install_ice_hook::{closure#1}>>::{closure#0}
   6:     0x7f99024279e2 - std[287527946f71f0bc]::panicking::panic_with_hook
   7:     0x7f990240a108 - std[287527946f71f0bc]::panicking::panic_handler::{closure#0}
   8:     0x7f9902400dc9 - std[287527946f71f0bc]::sys::backtrace::__rust_end_short_backtrace::<std[287527946f71f0bc]::panicking::panic_handler::{closure#0}, !>
   9:     0x7f990240bbdd - __rustc[5ec7081ce87b032f]::rust_begin_unwind
  10:     0x7f98fef0cfbc - core[7477a31a1a34cc49]::panicking::panic_fmt
  11:     0x7f990309c1fa - <rustc_infer[5a785d1c27494969]::infer::InferCtxt>::probe::<rustc_hir_typeck[ed5453188701a25b]::method::probe::ProbeResult, <rustc_hir_typeck[ed5453188701a25b]::method::probe::ProbeContext>::consider_probe::{closure#0}::{closure#0}>
  12:     0x7f990308b69d - <rustc_hir_typeck[ed5453188701a25b]::method::probe::ProbeContext>::pick_all_method
  13:     0x7f9902dc9ab6 - <rustc_hir_typeck[ed5453188701a25b]::fn_ctxt::FnCtxt>::lookup_probe
  14:     0x7f9902de6b41 - <rustc_hir_typeck[ed5453188701a25b]::fn_ctxt::FnCtxt>::check_expr_method_call
  15:     0x7f9902e19963 - <rustc_hir_typeck[ed5453188701a25b]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  16:     0x7f9902e14ac2 - <rustc_hir_typeck[ed5453188701a25b]::fn_ctxt::FnCtxt>::check_expr_block
  17:     0x7f9902e1998b - <rustc_hir_typeck[ed5453188701a25b]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  18:     0x7f9902e1b44b - <rustc_hir_typeck[ed5453188701a25b]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  19:     0x7f9902e13dbb - <rustc_hir_typeck[ed5453188701a25b]::fn_ctxt::FnCtxt>::check_expr_block
  20:     0x7f9902e1998b - <rustc_hir_typeck[ed5453188701a25b]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  21:     0x7f9902dfa1a6 - rustc_hir_typeck[ed5453188701a25b]::check::check_fn
  22:     0x7f9903795c20 - <rustc_hir_typeck[ed5453188701a25b]::fn_ctxt::FnCtxt>::check_expr_closure
  23:     0x7f9902e1b24b - <rustc_hir_typeck[ed5453188701a25b]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  24:     0x7f9902dfa1a6 - rustc_hir_typeck[ed5453188701a25b]::check::check_fn
  25:     0x7f99037b3d33 - rustc_hir_typeck[ed5453188701a25b]::typeck_with_inspect::{closure#0}
  26:     0x7f99037b26b4 - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::typeck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 8usize]>>
  27:     0x7f9902b5522a - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_data_structures[e0d8483206daafa2]::vec_cache::VecCache<rustc_span[851f35e07e3c2dd4]::def_id::LocalDefId, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[b7e73caa2c2ad9a]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  28:     0x7f9902b54b09 - rustc_query_impl[f20b8284200e680a]::query_impl::typeck::get_query_non_incr::__rust_end_short_backtrace
  29:     0x7f990305010a - rustc_mir_build[80b6bf0f2af09b99]::thir::pattern::check_match::check_match
  30:     0x7f990304f4d1 - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::check_match::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 1usize]>>
  31:     0x7f9903192f78 - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_data_structures[e0d8483206daafa2]::vec_cache::VecCache<rustc_span[851f35e07e3c2dd4]::def_id::LocalDefId, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 1usize]>, rustc_query_system[b7e73caa2c2ad9a]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  32:     0x7f9903192c3d - rustc_query_impl[f20b8284200e680a]::query_impl::check_match::get_query_non_incr::__rust_end_short_backtrace
  33:     0x7f9902f11667 - rustc_mir_build[80b6bf0f2af09b99]::builder::build_mir_inner_impl
  34:     0x7f9902a08dea - rustc_mir_transform[f63ee9052e691c]::mir_built
  35:     0x7f9902a08db3 - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::mir_built::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 8usize]>>
  36:     0x7f9902b5522a - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_data_structures[e0d8483206daafa2]::vec_cache::VecCache<rustc_span[851f35e07e3c2dd4]::def_id::LocalDefId, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[b7e73caa2c2ad9a]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  37:     0x7f9902b54c89 - rustc_query_impl[f20b8284200e680a]::query_impl::mir_built::get_query_non_incr::__rust_end_short_backtrace
  38:     0x7f9902a4eb97 - rustc_mir_transform[f63ee9052e691c]::ffi_unwind_calls::has_ffi_unwind_calls
  39:     0x7f9902a4e5a9 - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::has_ffi_unwind_calls::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 1usize]>>
  40:     0x7f9903192f78 - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_data_structures[e0d8483206daafa2]::vec_cache::VecCache<rustc_span[851f35e07e3c2dd4]::def_id::LocalDefId, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 1usize]>, rustc_query_system[b7e73caa2c2ad9a]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  41:     0x7f9903192afd - rustc_query_impl[f20b8284200e680a]::query_impl::has_ffi_unwind_calls::get_query_non_incr::__rust_end_short_backtrace
  42:     0x7f9903726bcd - rustc_mir_transform[f63ee9052e691c]::mir_promoted
  43:     0x7f99037269e0 - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::mir_promoted::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 16usize]>>
  44:     0x7f99033e9475 - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_data_structures[e0d8483206daafa2]::vec_cache::VecCache<rustc_span[851f35e07e3c2dd4]::def_id::LocalDefId, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 16usize]>, rustc_query_system[b7e73caa2c2ad9a]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  45:     0x7f99033e900c - rustc_query_impl[f20b8284200e680a]::query_impl::mir_promoted::get_query_non_incr::__rust_end_short_backtrace
  46:     0x7f9903b3d325 - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::mir_borrowck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 8usize]>>
  47:     0x7f9902b5522a - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_data_structures[e0d8483206daafa2]::vec_cache::VecCache<rustc_span[851f35e07e3c2dd4]::def_id::LocalDefId, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[b7e73caa2c2ad9a]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  48:     0x7f9902b4bf99 - rustc_query_impl[f20b8284200e680a]::query_impl::mir_borrowck::get_query_non_incr::__rust_end_short_backtrace
  49:     0x7f9903a143b5 - rustc_hir_analysis[1a80394817eddfdd]::collect::type_of::opaque::find_opaque_ty_constraints_for_rpit
  50:     0x7f9903a14042 - rustc_hir_analysis[1a80394817eddfdd]::collect::type_of::type_of_opaque
  51:     0x7f9903a13f47 - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::type_of_opaque::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 8usize]>>
  52:     0x7f9902a173c5 - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_query_system[b7e73caa2c2ad9a]::query::caches::DefIdCache<rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  53:     0x7f99033d90db - rustc_query_impl[f20b8284200e680a]::query_impl::type_of_opaque::get_query_non_incr::__rust_end_short_backtrace
  54:     0x7f99033bd8b6 - rustc_hir_analysis[1a80394817eddfdd]::collect::type_of::type_of
  55:     0x7f9902b75c26 - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::type_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 8usize]>>
  56:     0x7f9902a173c5 - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_query_system[b7e73caa2c2ad9a]::query::caches::DefIdCache<rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  57:     0x7f9902a16f6f - rustc_query_impl[f20b8284200e680a]::query_impl::type_of::get_query_non_incr::__rust_end_short_backtrace
  58:     0x7f9903a14abb - rustc_hir_analysis[1a80394817eddfdd]::check::check::check_opaque
  59:     0x7f990319b904 - rustc_hir_analysis[1a80394817eddfdd]::check::check::check_item_type
  60:     0x7f990319379c - rustc_hir_analysis[1a80394817eddfdd]::check::wfcheck::check_well_formed
  61:     0x7f990319376b - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::check_well_formed::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 1usize]>>
  62:     0x7f99031931e3 - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_data_structures[e0d8483206daafa2]::vec_cache::VecCache<rustc_span[851f35e07e3c2dd4]::def_id::LocalDefId, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 1usize]>, rustc_query_system[b7e73caa2c2ad9a]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  63:     0x7f9903192cd6 - rustc_query_impl[f20b8284200e680a]::query_impl::check_well_formed::get_query_non_incr::__rust_end_short_backtrace
  64:     0x7f99031904d1 - rustc_hir_analysis[1a80394817eddfdd]::check::wfcheck::check_type_wf
  65:     0x7f990319035d - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::check_type_wf::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 1usize]>>
  66:     0x7f9903ba3e85 - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_query_system[b7e73caa2c2ad9a]::query::caches::SingleCache<rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  67:     0x7f9903ba3b36 - rustc_query_impl[f20b8284200e680a]::query_impl::check_type_wf::get_query_non_incr::__rust_end_short_backtrace
  68:     0x7f990353f316 - rustc_hir_analysis[1a80394817eddfdd]::check_crate
  69:     0x7f9902b493db - rustc_interface[5c48368f5debe80d]::passes::analysis
  70:     0x7f9902b49035 - rustc_query_impl[f20b8284200e680a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f20b8284200e680a]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 0usize]>>
  71:     0x7f9903ba4be3 - rustc_query_system[b7e73caa2c2ad9a]::query::plumbing::try_execute_query::<rustc_query_impl[f20b8284200e680a]::DynamicConfig<rustc_query_system[b7e73caa2c2ad9a]::query::caches::SingleCache<rustc_middle[725063e074f29df]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[f20b8284200e680a]::plumbing::QueryCtxt, false>
  72:     0x7f9903ba49ce - rustc_query_impl[f20b8284200e680a]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  73:     0x7f9903d82c11 - <rustc_interface[5c48368f5debe80d]::passes::create_and_enter_global_ctxt<core[7477a31a1a34cc49]::option::Option<rustc_interface[5c48368f5debe80d]::queries::Linker>, rustc_driver_impl[517260a07841a630]::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core[7477a31a1a34cc49]::ops::function::FnOnce<(&rustc_session[eb79e6d3104f29ae]::session::Session, rustc_middle[725063e074f29df]::ty::context::CurrentGcx, alloc[979041e9952e771b]::sync::Arc<rustc_data_structures[e0d8483206daafa2]::jobserver::Proxy>, &std[287527946f71f0bc]::sync::once_lock::OnceLock<rustc_middle[725063e074f29df]::ty::context::GlobalCtxt>, &rustc_data_structures[e0d8483206daafa2]::sync::worker_local::WorkerLocal<rustc_middle[725063e074f29df]::arena::Arena>, &rustc_data_structures[e0d8483206daafa2]::sync::worker_local::WorkerLocal<rustc_hir[7eac7792dedc1861]::Arena>, rustc_driver_impl[517260a07841a630]::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  74:     0x7f9903bf5d34 - rustc_interface[5c48368f5debe80d]::interface::run_compiler::<(), rustc_driver_impl[517260a07841a630]::run_compiler::{closure#0}>::{closure#1}
  75:     0x7f9903b5aa3a - std[287527946f71f0bc]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[5c48368f5debe80d]::util::run_in_thread_with_globals<rustc_interface[5c48368f5debe80d]::util::run_in_thread_pool_with_globals<rustc_interface[5c48368f5debe80d]::interface::run_compiler<(), rustc_driver_impl[517260a07841a630]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  76:     0x7f9903b5a7fe - <std[287527946f71f0bc]::thread::lifecycle::spawn_unchecked<rustc_interface[5c48368f5debe80d]::util::run_in_thread_with_globals<rustc_interface[5c48368f5debe80d]::util::run_in_thread_pool_with_globals<rustc_interface[5c48368f5debe80d]::interface::run_compiler<(), rustc_driver_impl[517260a07841a630]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[7477a31a1a34cc49]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  77:     0x7f9903b5c5c4 - <std[287527946f71f0bc]::sys::thread::unix::Thread>::new::thread_start
  78:     0x7f98fd49698b - <unknown>
  79:     0x7f98fd51a9cc - <unknown>
  80:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.95.0-nightly (621d76794 2026-01-18) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z next-solver=globally -Z dump-mir-dir=dir

query stack during panic:
#0 [typeck] type-checking `real_main`
#1 [check_match] match-checking `real_main`
#2 [mir_built] building MIR for `real_main`
#3 [has_ffi_unwind_calls] checking if `real_main` contains FFI-unwind calls
#4 [mir_promoted] promoting constants in MIR for `real_main`
#5 [mir_borrowck] borrow-checking `real_main`
#6 [type_of_opaque] computing type of opaque `real_main::{opaque#0}`
#7 [type_of] computing type of `real_main::{opaque#0}`
#8 [check_well_formed] checking that `real_main::{opaque#0}` is well-formed
#9 [check_type_wf] checking that types are well-formed
#10 [analysis] running analysis passes on crate `mvce`
end of query stack
error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0601`.

```

</p>
</details>

@rustbot label +WG-trait-system-refactor

**Comment by matthiaskrgr** [comment]
Regression in nightly-2025-09-27
#146885

**Comment by clado** [comment]
hello! :)

I've been investigating this issue (I've never contributed to rust before but saw `E-needs-test` issues listed in the docs as a good way to learn 🙏) in the hopes of contributing a test, as requested/tagged. 🙇 

I did a git bisect and noticed that this ICE was resolved in #151676, specifically. That commit did include a regression test, which causes me to wonder whether an additional regression test is still needed here or whether #151676 fully resolves this issue. (I don't currently have a sufficient understanding of what cased these ICEs to make that determination myself. ^^;)

@adwinwhite as the author of the patch do you think that an additional regression test is potentially still needed, or can this issue be closed?
If the former, would a naive test provided+written by someone who's not fully sure how to [describe what exactly is being tested](https://rustc-dev-guide.rust-lang.org/tests/adding.html#comment-explaining-what-the-test-is-about) be useful? Or should I aim to develop that understanding for myself before contributing and/or pass this issue to someone who can? (Hoping to check in before I do too much more work on this. 🙇) 

For further context/the benefit of anyone working on this issue after me, the smallest test that I'd come up with was (based on matthiaskrgr's auto-reduction with some trivial(?) changes; error annotations left as an exercise for the reader ;) ):

```
//@ edition:2021
//@ compile-flags: -Znext-solver=globally

async fn recurse() {
    let mut vec: Vec<()> = recurse();
    vec.push(());
}
```

quite small, but remains relatively opaque as to what exactly is being tested...
